src/norm/postgres.nim

```diff
@@ -390,43 +390,42 @@ proc update*[T: Model](dbConn; obj: var T) =
 
   dbConn.exec(sql qry, row)
 
-proc update*[T: Model](dbConn; objs: var openArray[T]) =
-  ## Update rows for each `Model`_ instance in open array.
 
-  # for obj in objs.mitems:
-  #   dbConn.update(obj)
-
-  # return
-
-  type BulkUpdateValues = Table[string, seq[tuple[id: int64; val: DbValue]]]
-
-  func genBulkUpdateQuery(tableName: string; values: BulkUpdateValues): tuple[qry: string; values: seq[DbValue]] =
-    result.qry = fmt"UPDATE {tableName} SET{'\l'}"
-    var ids: seq[int64]
-    let limit = values.len - 1
-    var
-      i = -1
-      v = 1
-    for field, conds in values.pairs:
-      inc i
-      result.qry.add fmt"  {field} = CASE{'\l'}"
-      for (id, val) in conds:
-        result.qry.add fmt"    WHEN id = {id} THEN ${v}{'\l'}"
-        inc v
-        if id notin ids:
-          ids.add id
-        result.values.add val
-      result.qry.add "  END"
-      if i < limit:
-        result.qry.add ","
-      result.qry.add "\l"
-    let idsStr = ids.join ", "
-    result.qry.add fmt"WHERE id IN ({idsStr})"
-  proc add(t: var BulkUpdateValues; field: string; id: int64, value: DbValue) =
-    if not t.hasKey field:
-      t[field] = @[]
-    t[field].add (id, value)
+type BulkUpdateValues = Table[string, seq[tuple[id: int64; val: DbValue]]]
+  ## A table that stores all the data that will be changed in bulk query  
+  ## Just used by `genBulkUpdateQuery` func
+func genBulkUpdateQuery(tableName: string; values: BulkUpdateValues): tuple[qry: string; values: seq[DbValue]] =
+  ## Generates a single query for Postgres that modifies multiple columns of
+  ## multiple rows at same time
+  result.qry = fmt"UPDATE {tableName} SET{'\l'}"
+  var ids: seq[int64]
+  let limit = values.len - 1
+  var
+    i = 0
+    argNum = 1
+  for field, conds in values.pairs:
+    result.qry.add fmt"  {field} = CASE{'\l'}"
+    for (id, val) in conds:
+      result.qry.add fmt"    WHEN id = {id} THEN ${argNum}{'\l'}"
+      inc argNum
+      if id notin ids:
+        ids.add id
+      result.values.add val
+    result.qry.add "  END"
+    if i < limit:
+      result.qry.add ","
+    result.qry.add "\l"
+    inc i
+  let idsStr = ids.join ", "
+  result.qry.add fmt"WHERE id IN ({idsStr})"
+proc add(t: var BulkUpdateValues; field: string; id: int64, value: DbValue) =
```

Comment:
Please separate procs with blank lines.

In general, please make your code more sparse. It's much easier to read this way.

Reply:
ok